File: packages/shared/src/hmac.ts
/**
 * HMAC-SHA256 request signing for inter-service auth (Vercel Edge to FastAPI).
 *
 * ADR-0002 inter-service auth, ADR-0035 60s skew window.
 *
 * Canonical message (UTF-8): `${timestamp}.${nonce}.${bodySha256Hex}`
 *
 * The body is hashed once (SHA-256) so streaming bodies and large uploads do
 * not require keeping the entire bytes in memory twice. The Python verifier
 * uses the identical canonical form. See tests/golden_vectors.test.ts.
 *
 * Replay prevention: nonce stored in Redis SET on the API side with TTL equal
 * to (skew_seconds + safety_margin). A request whose timestamp is older than
 * skew_seconds is rejected before the nonce check.
 */

import {
  HMAC_HEADER_NONCE,
  HMAC_HEADER_SIGNATURE,
  HMAC_HEADER_TIMESTAMP,
  type HmacSignedRequestHeaders,
} from './contracts/index.js';

const DEFAULT_SKEW_SECONDS = 60;
const NONCE_BYTES = 16;

export interface SignRequestOptions {
  /** Shared HMAC secret (>= 32 chars). */
  secret: string;
  /** Request body bytes. Empty for GET. */
  body: Uint8Array | string;
  /** Override clock for tests. */
  nowMs?: number;
  /** Override nonce for tests (must be 16+ bytes base64url-encoded). */
  nonce?: string;
}

export interface VerifyRequestOptions {
  /** Shared HMAC secret. Also accepts the previous secret during rotation overlap. */
  secret: string;
  /** Optional previous secret during 24h rotation overlap window. */
  secretPrevious?: string | null;
  /** Body the API received. */
  body: Uint8Array | string;
  /** Headers from the incoming request. */
  signatureHeader: string;
  timestampHeader: string;
  nonceHeader: string;
  /** Reject if abs(now - timestamp) > skewSeconds. Default 60. */
  skewSeconds?: number;
  /** Override clock for tests. */
  nowMs?: number;
}

export interface VerifyResult {
  ok: boolean;
  reason?:
    | 'missing_header'
    | 'invalid_timestamp'
    | 'timestamp_skew_too_large'
    | 'signature_mismatch'
    | 'invalid_signature_format';
}

const HEX_LOWER = '0123456789abcdef' as const;

function bytesToHex(bytes: Uint8Array): string {
  const out = new Array<string>(bytes.length);
  for (let i = 0; i < bytes.length; i++) {
    const v = bytes[i] ?? 0;
    const hi = HEX_LOWER.charAt(v >> 4);
    const lo = HEX_LOWER.charAt(v & 0x0f);
    out[i] = hi + lo;
  }
  return out.join('');
}

function toBytes(input: Uint8Array | string): Uint8Array {
  if (typeof input === 'string') {
    return new TextEncoder().encode(input);
  }
  return input;
}

async function sha256Hex(input: Uint8Array): Promise<string> {
  const buf = await crypto.subtle.digest('SHA-256', input as unknown as ArrayBuffer);
  return bytesToHex(new Uint8Array(buf));
}

async function hmacSha256Hex(secret: string, message: string): Promise<string> {
  const key = await crypto.subtle.importKey(
    'raw',
    new TextEncoder().encode(secret),
    { name: 'HMAC', hash: 'SHA-256' },
    false,
    ['sign'],
  );
  const sig = await crypto.subtle.sign(
    'HMAC',
    key,
    new TextEncoder().encode(message) as unknown as ArrayBuffer,
  );
  return bytesToHex(new Uint8Array(sig));
}

/**
 * Build the canonical message: `${timestamp}.${nonce}.${sha256Hex(body)}`.
 */
export async function canonicalMessage(
  timestamp: string,
  nonce: string,
  body: Uint8Array | string,
): Promise<string> {
  const bodyHash = await sha256Hex(toBytes(body));
  return `${timestamp}.${nonce}.${bodyHash}`;
}

/**
 * Generate a cryptographically-strong nonce (base64url, no padding).
 */
export function generateNonce(): string {
  const bytes = new Uint8Array(NONCE_BYTES);
  crypto.getRandomValues(bytes);
  // base64url without padding
  let b64 = '';
  if (typeof btoa === 'function') {
    let bin = '';
    for (let i = 0; i < bytes.length; i++) {
      bin += String.fromCharCode(bytes[i] ?? 0);
    }
    b64 = btoa(bin);
  } else {
    // Node fallback
    b64 = Buffer.from(bytes).toString('base64');
  }
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

/**
 * Sign a request, returning the three headers the API will validate.
 */
export async function signRequest(options: SignRequestOptions): Promise<HmacSignedRequestHeaders> {
  if (options.secret.length < 32) {
    throw new Error('HMAC secret must be at least 32 characters.');
  }
  const nowMs = options.nowMs ?? Date.now();
  const timestamp = Math.floor(nowMs / 1000).toString();
  const nonce = options.nonce ?? generateNonce();
  const message = await canonicalMessage(timestamp, nonce, options.body);
  const signature = await hmacSha256Hex(options.secret, message);
  return {
    [HMAC_HEADER_SIGNATURE]: signature,
    [HMAC_HEADER_TIMESTAMP]: timestamp,
    [HMAC_HEADER_NONCE]: nonce,
  };
}

/**
 * Constant-time string compare. Returns false on length mismatch without leaking.
 */
function constantTimeEqual(a: string, b: string): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a.charCodeAt(i) ^ b.charCodeAt(i);
  }
  return diff === 0;
}

/**
 * Verify an incoming request signature. Does NOT check the nonce store; the
 * caller is responsible for the Redis SET replay check.
 */
export async function verifyRequest(options: VerifyRequestOptions): Promise<VerifyResult> {
  if (!options.signatureHeader || !options.timestampHeader || !options.nonceHeader) {
    return { ok: false, reason: 'missing_header' };
  }
  if (!/^[0-9a-f]{64}$/.test(options.signatureHeader)) {
    return { ok: false, reason: 'invalid_signature_format' };
  }
  const tsSeconds = Number.parseInt(options.timestampHeader, 10);
  if (!Number.isFinite(tsSeconds) || tsSeconds <= 0) {
    return { ok: false, reason: 'invalid_timestamp' };
  }
  const skewSeconds = options.skewSeconds ?? DEFAULT_SKEW_SECONDS;
  const nowSec = Math.floor((options.nowMs ?? Date.now()) / 1000);
  if (Math.abs(nowSec - tsSeconds) > skewSeconds) {
    return { ok: false, reason: 'timestamp_skew_too_large' };
  }
  const message = await canonicalMessage(
    options.timestampHeader,
    options.nonceHeader,
    options.body,
  );
  const expected = await hmacSha256Hex(options.secret, message);
  if (constantTimeEqual(expected, options.signatureHeader)) {
    return { ok: true };
  }
  if (options.secretPrevious) {
    const expectedPrev = await hmacSha256Hex(options.secretPrevious, message);
    if (constantTimeEqual(expectedPrev, options.signatureHeader)) {
      return { ok: true };
    }
  }
  return { ok: false, reason: 'signature_mismatch' };
}

File: apps/api-python/src/strota_api/security/hmac.py
"""HMAC-SHA256 request signing for inter-service auth.

The canonical message form is identical to the TypeScript implementation in
`packages/shared/src/hmac.ts`:

    f"{timestamp}.{nonce}.{sha256_hex(body)}"

Where `timestamp` is unix epoch seconds as a decimal string, `nonce` is a
base64url-encoded random value, and `sha256_hex(body)` is the lowercase hex
SHA-256 of the raw request body (empty string for GET).

Golden vectors in `packages/shared/tests/golden_vectors.test.ts` and the
matching pytest in `apps/api-python/tests/test_hmac_golden.py` lock the
TS and Python implementations to byte-identical output.
"""

from __future__ import annotations

import hashlib
import hmac as _hmac
import secrets
import time
from dataclasses import dataclass
from typing import Literal

HMAC_HEADER_SIGNATURE = "x-strota-signature"
HMAC_HEADER_TIMESTAMP = "x-strota-timestamp"
HMAC_HEADER_NONCE = "x-strota-nonce"

DEFAULT_SKEW_SECONDS = 60
NONCE_BYTES = 16
_SIGNATURE_HEX_RE = "0123456789abcdef"


VerifyReason = Literal[
    "missing_header",
    "invalid_timestamp",
    "timestamp_skew_too_large",
    "signature_mismatch",
    "invalid_signature_format",
]


@dataclass(frozen=True)
class VerifyOutcome:
    """Result of `verify_request`.

    `ok=True` means signature and timestamp are valid. Caller is responsible
    for the Redis nonce-replay check.
    """

    ok: bool
    reason: VerifyReason | None = None


def _sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def canonical_message(timestamp: str, nonce: str, body: bytes) -> str:
    """Build the canonical message: f'{timestamp}.{nonce}.{sha256_hex(body)}'."""
    return f"{timestamp}.{nonce}.{_sha256_hex(body)}"


def generate_nonce() -> str:
    """Generate a base64url nonce without padding (matches the TS implementation)."""
    raw = secrets.token_bytes(NONCE_BYTES)
    return _b64url_no_pad(raw)


def _b64url_no_pad(raw: bytes) -> str:
    import base64

    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def sign_request(
    *,
    secret: str,
    body: bytes,
    now_ms: int | None = None,
    nonce: str | None = None,
) -> dict[str, str]:
    """Compute the three HMAC headers for an outgoing request.

    Returns a dict with the lowercase header names so it can be passed directly
    to httpx.
    """
    if len(secret) < 32:
        msg = "HMAC secret must be at least 32 characters."
        raise ValueError(msg)
    if now_ms is None:
        now_ms = int(time.time() * 1000)
    timestamp = str(now_ms // 1000)
    nonce_value = nonce or generate_nonce()
    message = canonical_message(timestamp, nonce_value, body)
    signature = _hmac.new(
        secret.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    return {
        HMAC_HEADER_SIGNATURE: signature,
        HMAC_HEADER_TIMESTAMP: timestamp,
        HMAC_HEADER_NONCE: nonce_value,
    }


def _is_lowercase_hex_64(value: str) -> bool:
    if len(value) != 64:
        return False
    return all(c in _SIGNATURE_HEX_RE for c in value)


def verify_request(
    *,
    secret: str,
    body: bytes,
    signature_header: str,
    timestamp_header: str,
    nonce_header: str,
    skew_seconds: int = DEFAULT_SKEW_SECONDS,
    secret_previous: str | None = None,
    now_ms: int | None = None,
) -> VerifyOutcome:
    """Verify an incoming signature without consulting the nonce store.

    Returns `VerifyOutcome(ok=True)` when the signature is valid and the
    timestamp is within `skew_seconds` of `now`. The caller is responsible
    for the Redis nonce-SET replay check.
    """
    if not signature_header or not timestamp_header or not nonce_header:
        return VerifyOutcome(ok=False, reason="missing_header")
    if not _is_lowercase_hex_64(signature_header):
        return VerifyOutcome(ok=False, reason="invalid_signature_format")
    try:
        ts_seconds = int(timestamp_header)
    except ValueError:
        return VerifyOutcome(ok=False, reason="invalid_timestamp")
    if ts_seconds <= 0:
        return VerifyOutcome(ok=False, reason="invalid_timestamp")
    now_sec = (now_ms if now_ms is not None else int(time.time() * 1000)) // 1000
    if abs(now_sec - ts_seconds) > skew_seconds:
        return VerifyOutcome(ok=False, reason="timestamp_skew_too_large")

    message = canonical_message(timestamp_header, nonce_header, body)
    expected = _hmac.new(
        secret.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()
    if _hmac.compare_digest(expected, signature_header):
        return VerifyOutcome(ok=True)
    if secret_previous:
        expected_prev = _hmac.new(
            secret_previous.encode("utf-8"),
            message.encode("utf-8"),
            hashlib.sha256,
        ).hexdigest()
        if _hmac.compare_digest(expected_prev, signature_header):
            return VerifyOutcome(ok=True)
    return VerifyOutcome(ok=False, reason="signature_mismatch")
